source/opt/libspirv.cpp

```diff
@@ -16,91 +16,58 @@
-  reinterpret_cast<ir::IrLoader*>(builder)->AddInstruction(inst);
-  return SPV_SUCCESS;
-};
+SpvTools::SpvTools(spv_target_env env) : impl_(new Impl(env)) {}
 
-}  // annoymous namespace
+SpvTools::~SpvTools() {}
 
 void SpvTools::SetMessageConsumer(MessageConsumer consumer) {
-  SetContextMessageConsumer(context_, std::move(consumer));
+  SetContextMessageConsumer(impl_->context, std::move(consumer));
 }
 
-spv_result_t SpvTools::Assemble(const std::string& text,
-                                std::vector<uint32_t>* binary) {
+bool SpvTools::Assemble(const std::string& text,
+                        std::vector<uint32_t>* binary) const {
   spv_binary spvbinary = nullptr;
-  spv_diagnostic diagnostic = nullptr;
-
-  spv_result_t status = spvTextToBinary(context_, text.data(), text.size(),
-                                        &spvbinary, &diagnostic);
+  spv_result_t status = spvTextToBinary(impl_->context, text.data(),
+                                        text.size(), &spvbinary, nullptr);
   if (status == SPV_SUCCESS) {
     binary->assign(spvbinary->code, spvbinary->code + spvbinary->wordCount);
   }
-
-  spvDiagnosticDestroy(diagnostic);
   spvBinaryDestroy(spvbinary);
-
-  return status;
+  return status == SPV_SUCCESS;
 }
 
-spv_result_t SpvTools::Disassemble(const std::vector<uint32_t>& binary,
-                                   std::string* text, uint32_t options) {
+bool SpvTools::Disassemble(const std::vector<uint32_t>& binary,
+                           std::string* text, uint32_t options) const {
   spv_text spvtext = nullptr;
-  spv_diagnostic diagnostic = nullptr;
-
-  spv_result_t status = spvBinaryToText(context_, binary.data(), binary.size(),
-                                        options, &spvtext, &diagnostic);
+  spv_result_t status = spvBinaryToText(
+      impl_->context, binary.data(), binary.size(), options, &spvtext, nullptr);
   if (status == SPV_SUCCESS) {
     text->assign(spvtext->str, spvtext->str + spvtext->length);
   }
-
-  spvDiagnosticDestroy(diagnostic);
   spvTextDestroy(spvtext);
-
-  return status;
-}
-
-std::unique_ptr<ir::Module> SpvTools::BuildModule(
-    const std::vector<uint32_t>& binary) {
-  spv_diagnostic diagnostic = nullptr;
-
-  auto module = MakeUnique<ir::Module>();
-  ir::IrLoader loader(context_->consumer, module.get());
-
-  spv_result_t status =
-      spvBinaryParse(context_, &loader, binary.data(), binary.size(),
-                     SetSpvHeader, SetSpvInst, &diagnostic);
-
-  spvDiagnosticDestroy(diagnostic);
-
-  loader.EndModule();
-
-  if (status == SPV_SUCCESS) return module;
-  return nullptr;
+  return status == SPV_SUCCESS;
 }
 
-std::unique_ptr<ir::Module> SpvTools::BuildModule(const std::string& text) {
-  std::vector<uint32_t> binary;
-  if (Assemble(text, &binary) != SPV_SUCCESS) return nullptr;
-  return BuildModule(binary);
+bool SpvTools::Validate(const std::vector<uint32_t>& binary) const {
```

Comment:
My first impression is to be a little bit wary of losing information by only returning a bool.  For example, we've talked about succeeding but with warnings in case the validator sees extensions it doesn't understand.  But the message consumer already can express warnings that still allow you to succeed.  So Ok I think!
It just feels weird given that several Khronos APIs use 0 for success and other ints for failure.

But still ok.